docs/src/api/java.md

```diff
@@ -3,6 +3,206 @@
 
 Terminates this instance of Playwright, will also close all created browsers if they are still running.
 
+## method: BrowserContext.waitForPage
+* langs: java
+- returns: <[Page]>
+
+Runs callback and waits for a new Page.
+
+### option: BrowserContext.waitForPage.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: BrowserContext.waitForPage.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForClose
+* langs: java
+- returns: <[Page]>
+
+Runs callback and waits until the Page is closed.
+
+### option: Page.waitForClose.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForClose.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForConsole
+* langs: java
+- returns: <[ConsoleMessage]>
+
+Runs callback and waits for a console message.
+
+### option: Page.waitForConsole.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForConsole.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForDownload
+* langs: java
+- returns: <[Download]>
+
+Runs callback and waits for a download.
+
+### option: Page.waitForDownload.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForDownload.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForFileChooser
+* langs: java
+- returns: <[FileChooser]>
+
+Runs callback and waits for a file chooser.
+
+### option: Page.waitForFileChooser.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForFileChooser.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForFrameAttached
+* langs: java
+- returns: <[Frame]>
+
+Runs callback and waits for a [`event: Page.frameAttached`] event.
+
+### option: Page.waitForFrameAttached.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForFrameAttached.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForFrameDetached
+* langs: java
+- returns: <[Frame]>
+
+Runs callback and waits for a [`event: Page.frameDetached`] event.
+
+### option: Page.waitForFrameDetached.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForFrameDetached.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForFrameNavigated
+* langs: java
+- returns: <[Frame]>
+
+Runs callback and waits for navigation in one of the frames in the page.
+
+### option: Page.waitForFrameNavigated.timeout = %%-java-wait-for-event-timeout-%%
+
+### param: Page.waitForFrameNavigated.callback = %%-java-wait-for-event-callback-%%
+
+## method: Page.waitForPageError
```

Comment:
Maybe we don't need this one?
Also, I am not sure about waitForFrameAttached, waitForFrameDetached, waitForRequestFailed/Finished, waitForSocketError.

Reply:
Well, it is actually used in `TestWorker.shouldReportErrors` test today, without it the test would be more cumbersome. Same with `waitForRequestFailed`. `waitForFrameAttached` does as much sense as the `frameAttached` event itself, so I'd leave it there.

What is your criteria for leaving/removing some of these methods?

Reply:
Ok, I reconsidered this. The error condition should be exceptions rather then something that people put into the expectations and we already have `waitForRequest/Response` methods so adding `waitForRequestFailed` would just confuse people. There are no tests right now that use wait for frame attached/detached, so removing those too. Let's add them if there if there our clients ask for them.